### ISSUE TITLE
`String#gsub` alters invalid UTF-8 bytes

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#12547]
When replacing a character in a string with `gsub` it may alter the resulting string in unintended ways.

Bytes that are not a valid UTF-8 codepoint should be unaffected by `gsub`. But they may get replaced by the Unicode replacement character (U+FFFD).
The conditions are a single character (either represented as Char or String) being replaced by a string with more than one character or a single non-ASCII character.

```cr
"\x8B:".gsub(":", "..") # => "\uFFFD.."
# In contrast, if replaced by a single character:
"\x8B:".gsub(":", ".")  # => "\x8B."
```

The cause for this is that `String#gsub` iterates the string, decoding each character in order to compare it with the search character. If it does not match, the original data is used. 

https://github.com/crystal-lang/crystal/blob/12a9b5fc63b3285cc55e21d84705736d0c2ad4e2/src/string.cr#L2609

However, it's not the original data because the decoding mechanism yields the replacement character for invalid bytes.

I think this shouldn't be hard to fix. But then I'm also wondering if there might be similar cases like this in stdlib where char decoding has some unwanted effect like this one.

**Comment by straight-shoota** [comment]
For a complete solution, `#gsub(&)` could yield the current byte as second block parameter when it's not a valid codepoint. As a consequence, the same would probably apply to `#each_char` and `Char::Reader#each`.

**Comment by HertzDevil** [comment]
> Bytes that are not a valid UTF-8 codepoint should be unaffected by `gsub`. But they may get replaced by the Unicode replacement character (U+FFFD).

I don't see why this is an issue as long as we aren't mutating the original string. In fact methods like `#reverse` require invalid bytes to be mapped to U+FFFD for them to behave correctly (#12132).

**Comment by HertzDevil** [comment]
Rather it is the replacement argument that gets transformed in a wrong way:

```crystal
"\xC3.".gsub('.', "\x80") # => "\u{FFFD}\u{80}"
```

**Comment by straight-shoota** [comment]
> I don't see why this is an issue

If I ask to replace a specific character in a string, I don't want that to affect any other content of that string which is not that character.
Also it's a problem of consistency. In most circumstances, `#gsub` does *not* transform invalid bytes. This is an issue with a specific implementation strategy which applies on some input parameter conditions.

**Comment by HertzDevil** [comment]
> I don't want that to affect any other content of that string which is not that character.

Exactly; this is why the `sub` example should _not_ produce a valid byte sequence from invalid byte sequences from two different strings. And naturally `gsub` should behave identically to `sub` if there is only one occurrence of the replacement string.

> Also it's a problem of consistency.

Making `#gsub` always replace invalid byte sequences (for example via `String.build`) also provides consistency.

**Comment by asterite** [comment]
If in doubt, we can check what Ruby does. And do that, as Ruby usually does the right thing :-)

**Comment by HertzDevil** [comment]
Ruby always preserves the byte contents of a `String` because their strings are mutable and even their encoding can change. Therefore their behavior does not apply to Crystal's strings.

**Comment by asterite** [comment]
Given that a Crystal string doesn't necessarily represent a valid UTF-8 string (like, it can contain arbitrary binary data) I wouldn't expect `gsub` to change something I didn't ask. Just my opinion, though.

**Comment by HertzDevil** [comment]
> I wouldn't expect `gsub` to change something I didn't ask.

This can be taken to mean either the bytes or the characters. I prefer the latter because `Char`-based access is the predominant API for `String`.
